### PR TITLE
Add more date validation to activities program date

### PR DESCRIPTION
**Pull request by coltborg**
#### 🔗 Jira ticket
[CCAP-21](https://codeforamerica.atlassian.net/browse/CCAP-21)

#### ✍️ Description
This PR is to [address comments here](https://codeforamerica.atlassian.net/browse/CCAP-21?focusedCommentId=10030) in regards to date validation.

- Add validation if letters are provided
- Add validation if a date is before `01/01/1901`
- Add validation if a date is after `today`
- Expand tests to include these new validations
- `activityProgram` dates and `partnerProgram` dates have the same validation

#### ✅ Completion tasks

- [x] Added relevant tests
- [x] Meets acceptance criteria

[CCAP-21]: https://codeforamerica.atlassian.net/browse/CCAP-21?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgram.java [hunk 1]
```diff
@@ -17,43 +16,82 @@
...
     @Override
     public Map<String, List<String>> runValidation(FormSubmission formSubmission, Submission submission) {
 
         Locale locale = LocaleContextHolder.getLocale();
         Map<String, List<String>> errorMessages = new HashMap<>();
         Map<String, Object> inputData = formSubmission.getFormData();
 
-        String startMonth = inputData.get(INPUT_NAME_START_MONTH).toString();
-        String startYear = inputData.get(INPUT_NAME_START_YEAR).toString();
+        String startGroup = "activitiesProgramStart";
```
Comment (rapicastillo) [inline src/main/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgram.java:31]: Could we define this as a constant?

File: src/main/resources/messages.properties [hunk 2]
```diff
@@ -91,6 +91,7 @@ errors.required-financial-assistance=Please answer whether you need child care a
 errors.invalid-date-format=Make sure the date you entered is in this format: mm/dd/yyyy
 errors.invalid-birthdate-format=Make sure the birthdate you entered is in this format: mm/dd/yyyy
 errors.invalid-birthdate-range=Make sure the date you entered is between 01/01/1901 and today.
+errors.invalid-date-range=Make sure the date you entered is after 01/01/1901 and before today.
```
Comment (rapicastillo) [inline src/main/resources/messages.properties:94]: Feels like the error message is the same as `errors.invalid-birthdate-range` but is worded weirdly. Should we have the same verbiage as `errors.invalid-birthdate-range` ?

File: src/test/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgramTest.java [hunk 3]
```diff
@@ -85,13 +121,114 @@ void shouldErrorWhenEndMonthIsPresentAndEndYearIsMissing() {
...
+                "activitiesProgramStartYear", "1700",
+                "activitiesProgramEndMonth", "01",
+                "activitiesProgramEndDay", "",
+                "activitiesProgramEndYear", "2024"
+        );
+        FormSubmission submission = new FormSubmission(formData);
+        Map<String, List<String>> errors = validator.runValidation(submission, new Submission());
+        assertThat(errors).containsKey("activitiesProgramStart");
+        assertThat(errors.get("activitiesProgramStart")).contains("Make sure the date you entered is after 01/01/1901 and before today.");
+    }
```
Comment (rapicastillo) [inline src/test/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgramTest.java:183]: Hey! (Pardon if this is nitpicky) but I tried 
```
   ....
    "activitiesProgramStartMonth", "01",
    "activitiesProgramStartDay", "",
    "activitiesProgramStartYear", "1901",
   ...
```

Not sure if this is a bug but this still throws an error (setting end date for the date today does not throw an error, so I was assuming that this is an inclusive between)

**Review by rapicastillo — COMMENTED** [review]
I'll defer to what @enyia21 says but I left some questions and messages. Thanks Cypress!

File: src/main/resources/messages.properties (hunk 2 above)
Comment (coltborg) [reply src/main/resources/messages.properties:94]: Ya, I can align them.

**Review by enyia21 — CHANGES_REQUESTED** [review]

File: src/test/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgramTest.java (hunk 3 above)
Comment (coltborg) [reply src/test/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgramTest.java:183]: The logic on `01/1901` being inclusive or not is in [the VerifyDate action we're extending](https://github.com/codeforamerica/il-gcc-form-flow/blob/7d8dc7b0c623e43db1536cad84699ca3d6a0fe60/src/main/java/org/ilgcc/app/submission/actions/VerifyDate.java#L21-L28). 


And seeing Carl's message about this in [our stand up zoom](https://us06web.zoom.us/launch/chat/v2/eyJtaWQiOiI3LTExNkQ1N0FBLTEzMkQtNDNFQy1CRTk3LTZBOTg0QTA1RTY2RSIsInNpZCI6IjZjNzdhMjI3OTEyNTQ4OWE5MTI1NDllMDY0N2ZmZGY2QGNvbmZlcmVuY2UueG1wcC56b29tLnVzIiwidGltZSI6MTcxNDQwNzI0NzcxOX0), I'm going to keep it the same for now as this work didn't change the existing behavior.

> "Let’s not worry about it too much. Not many of our dates will actually be on 1/1/1901"

File: src/main/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgram.java [hunk 4]
```diff
@@ -17,43 +16,82 @@
...
+        String startDayField = startGroup.concat("Day");
+        String startYearField = startGroup.concat("Year");
+        String startMonth = inputData.get(startMonthField).toString();
+        String startDay = inputData.get(startDayField).toString().isEmpty() ? "01" : inputData.get(startDayField).toString();
+        String startYear = inputData.get(startYearField).toString();
+        String startDate = String.format("%s/%s/%s", startMonth, startDay, startYear);
 
-        String endMonth = inputData.get(INPUT_NAME_END_MONTH).toString();
-        String endYear = inputData.get(INPUT_NAME_END_YEAR).toString();
+        String endGroup = "activitiesProgramEnd";
```
[the remark this answers is not in the document]
Comment (coltborg) [reply src/main/java/org/ilgcc/app/submission/actions/ValidateMonthYearActivitiesProgram.java:40]: Yes, looking into this now

**Comment by coltborg** [comment]
@rapicastillo / @enyia21, I've made [changes to my PR to address](https://github.com/codeforamerica/il-gcc-form-flow/pull/219/commits/6c15b017344f53aef3aedad4794792033afe536c) the messaging and DRY up the code

**Review by enyia21 — APPROVED** [review]

**Pull request merged**